### PR TITLE
Auth implicit flow

### PR DESCRIPTION
**Pull request by ChrisTerBeke**
This PR is built on the awesome work in https://github.com/ChrisTerBeke/ThingiBrowser/pull/49 and replaces the current static authentication with an OAuth2 implicit flow. This is more future proof and allows users to sign in with their own Thingiverse or MyMiniFactory account and see their own things.

### CODE REVIEW

File: ThingiBrowser/drivers/myminifactory/MyMiniFactoryApiClient.py [hunk 1]
```diff
@@ -1,39 +1,77 @@
...
-from ...api.JsonObject import ApiError, Collection, Thing, ThingFile
+from ...api.JsonObject import ApiError, Collection, Thing, ThingFile, UserData
+from ...api.LocalAuthService import LocalAuthService
 
 
 class MyMiniFactoryApiClient(AbstractApiClient):
     """ Client for interacting with the MyMiniFactory API. """
 
     def __init__(self) -> None:
+        self._username: Optional[str] = None
```
Comment (BohunkG4mer) [inline ThingiBrowser/drivers/myminifactory/MyMiniFactoryApiClient.py:22]: Did you mean `self._username = None  # type: Optional[str]`? This line currently causes the plugin to fail while being loaded in Cura.

Comment (ChrisTerBeke) [reply ThingiBrowser/drivers/myminifactory/MyMiniFactoryApiClient.py:22]: You're right. Cura is using Python 3.5 in the built distribution, but I'm running 3.7 while running Cura from source, so I didn't catch this.

File: ThingiBrowser/api/LocalAuthService.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    def _onTokenReceived(self, token: Optional[str] = None) -> None:
+        """
+        Handler for when an implicit auth token was received.
+        Closes the server so the port can be re-used.
+        :param token: The received auth token.
+        """
+        if self._token_callback:
+            self._token_callback(token)
+        self._server.server_close()
```
Comment (BohunkG4mer) [inline ThingiBrowser/api/LocalAuthService.py:28]: Needs to be

```
self._server.shutdown()
self._server.server_close()
```

**Comment by BohunkG4mer** [comment]
Also, I see that on the Thingiverse app I created for this, it says once I set it to published, 10 people can use it until they finish reviewing it and approving it thenselves. So there may be some lag time there before this can be released. It also requires some contact info so it might be best for you to create a new one with your info.

**Comment by ChrisTerBeke** [comment]
@BohunkG4mer I'm looking into that Qt threading issue but so far cannot reproduce it (on Linux).
In the meantime I'm creating new credentials. Unfortunately the developer portal of MyMiniFactory is not working well and throws a health-check API error every few calls (so rarely I can actually get to the "api keys" screen and create a new key).

**Comment by ChrisTerBeke** [comment]
@BohunkG4mer I figured out what was going on with that crash you got. Basically Qt wants any invocation of methods marked with PyQt.... to happen in the Qt thread. Because of the callbacks we had in place to pass the token from the web server's request handler (which ran in a separate thread) this caused a chain of methods (and eventually some PyQt... marked methods on the ThingiService) to be called in that separate thread. I have replaced all usages of callbacks with Signals (which can relay message across threads safely) and I can now repeatedly login to both services without issue. Right now I never stop the web server or join the thread because that hangs the main thread (e.g. the whole of Cura) for a while, resulting in poor UX. This is something we can improve later. Please test some more and then I think it's ready to be merged. I'll wait with releasing until the credentials for both sites are in order.

**Comment by BohunkG4mer** [comment]
> @BohunkG4mer I'm looking into that Qt threading issue but so far cannot reproduce it (on Linux).
> In the meantime I'm creating new credentials. Unfortunately the developer portal of MyMiniFactory is not working well and throws a health-check API error every few calls (so rarely I can actually get to the "api keys" screen and create a new key).

Yeah, it's broken and I had to reach out to get mine. You shouldn't need an API key generated now that we are gunna do OAuth2 though. Just need the new client ID.

**Comment by BohunkG4mer** [comment]
> Please test some more and then I think it's ready to be merged. I'll wait with releasing until the credentials for both sites are in order.

Okay, sounds good. I'll grab the latest and test it out tonight.

**Comment by ChrisTerBeke** [comment]
> > @BohunkG4mer I'm looking into that Qt threading issue but so far cannot reproduce it (on Linux).
> > In the meantime I'm creating new credentials. Unfortunately the developer portal of MyMiniFactory is not working well and throws a health-check API error every few calls (so rarely I can actually get to the "api keys" screen and create a new key).
> 
> Yeah, it's broken and I had to reach out to get mine. You shouldn't need an API key generated now that we are gunna do OAuth2 though. Just need the new client ID.

I was planning on keeping the hardcoded tokens in there as a fallback so that we don't force all users to sign in with an account (they might not have) just to get the public models (like popular). I have introduced that behavior in https://github.com/ChrisTerBeke/ThingiBrowser/pull/50/commits/41b45effb0d66cef7563cd4eecf9421187b2d18b.

**Comment by BohunkG4mer** [comment]
> I was planning on keeping the hardcoded tokens in there as a fallback so that we don't force all users to sign in with an account (they might not have) just to get the public models (like popular). I have introduced that behavior in https://github.com/ChrisTerBeke/ThingiBrowser/pull/50/commits/41b45effb0d66cef7563cd4eecf9421187b2d18b.

Oh, smart. Then yeah, you have to reach out to MyMiniFactory support to get thag generated for you.

**Pull request merged**